datafusion/sql/src/unparser/expr.rs

```diff
@@ -113,10 +113,38 @@ impl Unparser<'_> {
             }
             Expr::Case(Case {
                 expr,
-                when_then_expr: _,
-                else_expr: _,
+                when_then_expr,
+                else_expr,
             }) => {
-                not_impl_err!("Unsupported expression: {expr:?}")
+                let conditions = when_then_expr
+                    .iter()
+                    .map(|(w, _)| self.expr_to_sql(w))
+                    .collect::<Result<Vec<_>>>()?;
+                let results = when_then_expr
+                    .iter()
+                    .map(|(_, t)| self.expr_to_sql(t))
+                    .collect::<Result<Vec<_>>>()?;
```

Comment:
I'm trying to make these two iterations to one, but I failed. ☹️
I tried
1. `unzip`, but I got two vec which consist of `Result<Expr>` but not `Expr`
2. `match`, got type incompatible error 🥲
3. `collect`, but don't know how to collect a `(Result<Expr>, Result<Expr>)`

Reply:
I likewise tried to use `izip` and came up with this, which wasn't particularly nice

```rust
                let (conditions, results): (Vec<_>, Vec<_>) = when_then_expr
                    .iter()
                    .map(|(w, t)| (self.expr_to_sql(w), self.expr_to_sql(t)))
                    .unzip();

                let conditions = conditions.into_iter().collect::<Result<Vec<_>>>()?;
                let results = results.into_iter().collect::<Result<Vec<_>>>()?;
```

This is the best I could come up with but I am not sure how much better it is:

```rust
                let mut conditions = vec![];
                let mut results = vec[]!;
                for (w, t) in when_then_expr {
                    conditions.push(self.expr_to_sql(w)?);
                    results.push(self.expr_to_sql(t)?);
                }
```

Reply:
Yeah, I also tried two pushes before but I removed it for declaring two mut vec 😆